### ISSUE TITLE
ERROR: failed to get previous image: parsing auth header '': unknown auth type from header: 

### ISSUE THREAD

**Issue by dmitry-mightydevops** [buildpacks/lifecycle#651]
### Summary

When I'm trying to pull image using ` /cnb/lifecycle/creator` it fails with the error:

```
ERROR: failed to get previous image: parsing auth header '': unknown auth type from header: 
```

##### Current behavior

```bash
export TOKEN=$(az acr login --name project --expose-token --output tsv --query accessToken --resource-group ProjectNameRG)
WARNING: Argument 'resource_group_name' has been deprecated and will be removed in a future release.

docker login reponame.azurecr.io --username 00000000-0000-0000-0000-000000000000 --password $TOKEN
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
Login Succeeded

cnb@project-backend-prod-build-pipeline-run-bkzcw-buildpack-xc6w-fb2nx:/layers$ /cnb/lifecycle/creator \
>         -app=/workspace/app/ \
>         -project-metadata=/workspace/app/project.toml \
>         -cache-dir=/cache \
>         -layers=/layers \
>         -platform=/platform \
>         -report=/layers/report.toml \
>         -cache-image= \
>         -uid=1000 \
>         -gid=1000 \
>         -process-type=web \
>         -skip-restore=false \
>         -previous-image=project.azurecr.io/project/backend:prod-9c93f94 \
>         -run-image=project.azurecr.io/buildpack-runner:v1 \
>         project.azurecr.io/project/backend:prod-9c93f94
===> DETECTING
3 of 4 buildpacks participating
google.python.runtime    0.9.1
google.config.entrypoint 0.9.0
google.utils.label       0.0.1
===> ANALYZING
ERROR: failed to get previous image: parsing auth header '': unknown auth type from header: 
cnb@project-backend-prod-build-pipeline-run-bkzcw-buildpack-xc6w-fb2nx:/layers$
```

##### Expected

For the image to be built and pushed to AZ ACR

---

##### lifecycle version

```
Lifecycle:
  Version: 0.11.1
  Buildpack APIs:
    Deprecated: (none)
    Supported: 0.2, 0.3, 0.4, 0.5, 0.6
  Platform APIs:
    Deprecated: (none)
    Supported: 0.3, 0.4, 0.5, 0.6
```
##### platform version(s)

```bash
➜ pack report                                                  
Pack:
  Version:  0.18.1+git-b5c1a96.build-2373
  OS/Arch:  linux/amd64

Default Lifecycle Version:  0.10.2

Supported Platform APIs:  0.3, 0.4

➜ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
```

##### anything else?

tekton info
```
➜ tkn version                                                        
Client version: 0.18.0
Pipeline version: v0.24.2
Triggers version: v0.14.2
Dashboard version: v0.17.0
```

tekton task step
```bash
- image: $(params.builder_image)
    imagePullPolicy: Always
    name: build
    resources: {}
    script: |
      #!/bin/bash
      az login --identity --username $AZURE_MSI_CLIENT_ID
      export TOKEN=$(az acr login --name project --expose-token --output tsv --query accessToken)
      docker login project.azurecr.io --username 00000000-0000-0000-0000-000000000000 --password $TOKEN

      # build image utilizing backpack
      /cnb/lifecycle/creator \
        -app=/workspace/app/$(params.source_subpath) \
        -project-metadata=/workspace/app/project.toml \
        -cache-dir=/cache \
        -layers=/layers \
        -platform=/platform \
        -report=/layers/report.toml \
        -cache-image=$(params.cache_image) \
        -uid=$(params.user_id) \
        -gid=$(params.group_id) \
        -process-type=$(params.process_type) \
        -skip-restore=$(params.skip_restore) \
        -previous-image=$(resources.outputs.image.url) \
        -run-image=$(params.run_image) \
        $(resources.outputs.image.url)
    securityContext:
      runAsGroup: 1000
      runAsUser: 1000
    volumeMounts:
    - mountPath: /layers
      name: layers-dir
    - mountPath: /cache
      name: $(params.cache)
    - mountPath: /platform
      name: $(params.platform_dir)
  volumes:
  - emptyDir: {}
    name: empty-dir
  - emptyDir: {}
    name: layers-dir
  workspaces:
  - name: source
```

**Comment by yaelharel** [comment]
Hi @dmitry-mightydevops, thank you for reporting this bug.
We're taking a look at this.

A small question:
Is there any chance you ran the same scenario on an older version of lifecycle?

Thanks!

**Comment by yaelharel** [comment]
Also, I saw your comment https://github.com/buildpacks/lifecycle/issues/213#issuecomment-874408044 and that we weren't able to reproduce this issue the last time it happened.

If you run the same scenario again, do you get the same error on every run?

Thanks!

**Comment by yaelharel** [comment]
**The bottom line:**
We might need to add an if condition to `authConfigToHeader` to check the value of `config.IdentityToken` in the `env_keychain` file: https://github.com/buildpacks/lifecycle/blob/f15f80c6731ca9b01545eb8234281a105d3f4577/auth/env_keychain.go#L143-L159

The error message is:
```
===> ANALYZING
ERROR: failed to get previous image: parsing auth header '': unknown auth type from header: 
cnb@project-backend-prod-build-pipeline-run-bkzcw-buildpack-xc6w-fb2nx:/layers$
```
We get it from the following flow:
* [analyzeArgs::analyze](https://github.com/buildpacks/lifecycle/blob/f15f80c6731ca9b01545eb8234281a105d3f4577/cmd/lifecycle/analyzer.go#L184)
* imgutil - [NewImage](https://github.com/buildpacks/imgutil/blob/8ba00079c71c25e65fdf7c794802c8b34ebc0644/remote/remote.go#L73)
* imgutil - [processPreviousImageOption](https://github.com/buildpacks/imgutil/blob/8ba00079c71c25e65fdf7c794802c8b34ebc0644/remote/remote.go#L122)
* imgutil - [newV1Image](https://github.com/buildpacks/imgutil/blob/8ba00079c71c25e65fdf7c794802c8b34ebc0644/remote/remote.go#L179)
* imgutil - [referneceForRepoName](https://github.com/buildpacks/imgutil/blob/8ba00079c71c25e65fdf7c794802c8b34ebc0644/remote/remote.go#L237)
* [ResolvedKeychain::Resolve](https://github.com/buildpacks/lifecycle/blob/f15f80c6731ca9b01545eb8234281a105d3f4577/auth/env_keychain.go#L57)
* [authHeaderToConfig](https://github.com/buildpacks/lifecycle/blob/f15f80c6731ca9b01545eb8234281a105d3f4577/auth/env_keychain.go#L166)

We can see from the error message that the header exists but it’s an empty string.
That means that we set it to an empty string in [InMemoryKeychain](https://github.com/buildpacks/lifecycle/blob/f15f80c6731ca9b01545eb8234281a105d3f4577/auth/env_keychain.go#L51).
This happens when [authConfigToHeader](https://github.com/buildpacks/lifecycle/blob/f15f80c6731ca9b01545eb8234281a105d3f4577/auth/env_keychain.go#L143) returns an empty string.
When looking at the last function, we can see that we ignore the `config.IdentityToken` value.

In `ggcr`: [defaultKeychain::Resolve](https://github.com/google/go-containerregistry/blob/b448abac9a7021605f2142ed6e04daa61c769324/pkg/authn/keychain.go#L62) we call to `cf.GetAuthConfig` where `cf` is `*configfile.ConfigFile` of `Docker` and we can see in `docker` [native_store::getCredentialsFromStore](https://github.com/docker/cli/blob/5f07d7d5a12423c0bc1fb507f4d006ad0cdfef42/cli/config/credentials/native_store.go#L129) that it’s setting the `IdentityToken` to `creds.Secret`.

So although the value of `IdentityToken` is set, `lifecycle` is ignoring it.
According to the "tekton task step", the beginning of the script is:
```
#!/bin/bash
az login --identity --username $AZURE_MSI_CLIENT_ID
export TOKEN=$(az acr login --name project --expose-token --output tsv --query accessToken)
docker login project.azurecr.io --username 00000000-0000-0000-0000-000000000000 --password $TOKEN
```
and we can clearly see that the `TOKEN` was set.

Next steps:
* Reproduce the bug.
* Work on a fix (and verify that the assumption is correct).

**Comment by yaelharel** [comment]
Thinking about this again (thank you @natalieparellano), maybe the problem relates to [this line](https://github.com/buildpacks/lifecycle/blob/f15f80c6731ca9b01545eb8234281a105d3f4577/auth/env_keychain.go#L152), where we check that both the `config.Username` and `config.Password` aren't empty.
According to the following line in the bug description, the username is `00000000-0000-0000-0000-000000000000`:
```
docker login reponame.azurecr.io --username 00000000-0000-0000-0000-000000000000 --password $TOKEN
```
Maybe from some reason, this username is being translated to an empty string (although it looks like the line above follows the suggestion [here](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-authentication?tabs=azure-cli#az-acr-login-with---expose-token))?

We would love to know whether this bug happens on every run, or it was a one time failure, because it may give us a hint on the reason of this failure.

**Comment by yaelharel** [comment]
Following a discussion that we had in the [ggcr issue](https://github.com/google/go-containerregistry/issues/1078) that I opened, we think that the problem is that we're ignoring the `IdentityToken` in our code.
@dmitry-mightydevops, can you please check if the `identitytoken` is set in your config file? Also, can you please let us know if you experience this failure on every run of the scenario you described?

**Comment by jromero** [comment]
FWIW, I just ran into this issue. 

My case had nothing to do with `IdentityToken` but with the fact that we had some configuration set to basic auth but with empty username and password (intended to be `Anonymous`). 

This section of code isn't returning an error which means the anything calling this would think things are ok when in reality they aren't.

https://github.com/buildpacks/lifecycle/blob/8f801a55fd4a1c9ee12eadd0b1c5b5cbf5ed98b5/auth/env_keychain.go#L144-L159

The reproductions steps should be fairly straight forward given what I just hit:

1. Set the docker config with a basic auth for `gcr.io/<project>/` with empty username and password
2. Call lifecycle `analyzer` with an image on repo `gcr.io/<project>/`


#### Configuration (kpack)

```
Name:         docker-secret
Namespace:    test
Labels:       <none>
Annotations:  kpack.io/docker: a7a6924fdb7b.ngrok.io

Type:  kubernetes.io/basic-auth

Data
====
password:  0 bytes
username:  0 bytes
```

**Comment by natalieparellano** [comment]
It looks like there are two issues here:

> we're ignoring the `IdentityToken` in our code

@dmitry-mightydevops I believe this is the cause of your issue. Is this issue still of concern to you? From what I recall, the work stalled out on this because of difficulty procuring an Azure environment to test with. If you (or someone with an Azure environment) would be willing to try out a dev version of the lifecycle in order to validate the changes, that might unblock this work.

> This section of code isn't returning an error which means the anything calling this would think things are ok when in reality they aren't.

This seems relatively straightforward to solve. Maybe we could make a new issue for it, since it's unrelated to the first one. Then again, @jromero is this issue still of concern to you? 

It would be good to have a sense if either of the issues mentioned are still of concern, given the time that has gone by.

**Comment by jromero** [comment]
Not concerning yet a poor experience.

May I propose a good first issue label given it's low complexity in case anyone would like to take this on?

**Comment by jjbustamante** [comment]
I faced this issue today trying to test my fix for ticket #830, let me summarize my findings.

- I have a private gcr.io registry where I am trying to publish an image using pack
- Locally I am running pack + custom lifecycle image (the one modified to test the issue)

The local execution is done with the following command

```shell
pack build gcr.io/cnb-playground/test-ruby-app --path ./ruby-sample-app \ 
    --cache-image gcr.io/cnb-playground/ruby-sample-app-cache  \ 
    --lifecycle-image lifecycle-layout --publish \ 
    --buildpack ./ruby-buildpack --verbose
```
I got the following error:

```shell
[exporter] error is processBaseImageOption... index.docker.io/cnbs/sample-stack-run@sha256:7b0946aa031f1ddd30bda8963eb33431458a25a576208646348ce89536fef59a
[exporter] ERROR: failed to create new app image: parsing auth header: unknown auth type from header
ERROR: failed to build: executing lifecycle. This may be the result of using an untrusted builder: 
failed with status code: 1
```

Which is similar to the error describe in this issue but in my case it happened in the **exporter** phase. The first line in the log was added my me to see which image was throwing the error, as we can see it is trying to find for the run-image in docker. `index.docker.io/cnbs/sample-stack-run@sha256:7b0946aa031f1ddd30bda8963eb33431458a25a576208646348ce89536fef59a`

Checking a little bit more the log, I noticed the following:

```shell
===> EXPORTING
Running the exporter on OS linux with:
Container Settings:
  Args: /cnb/lifecycle/exporter -log-level debug -app /workspace -cache-dir /cache -stack /layers/stack.toml -cache-image gcr.io/cnb-playground/ruby-sample-app-cache:latest gcr.io/cnb-playground/test-ruby-app
  System Envs: CNB_USER_ID=1000 CNB_GROUP_ID=1000 CNB_REGISTRY_AUTH={"gcr.io":"Basic X...FBYw==","index.docker.io":""} CNB_PLATFORM_API=0.8
  Image: index.docker.io/library/lifecycle-layout:latest
  User: root
  Labels: map[author:pack]
Host Settings:
  Binds: pack-layers-klqkgelzvs:/layers pack-app-qtquxyvuyy:/workspace
  Network Mode:
```

From the log, we can see the value of the environment variable `CNB_REGISTRY_AUTH={"gcr.io":"Basic X...FBYw==","index.docker.io":""}`, my gcr.io repo has a basic authentication token, but for docker is empty. In order to fix it I login to docker using the CLI tool `docker login -u [myusername]` I ran again the command and in that case the environment variable had the following: `REGISTRY_AUTH={"gcr.io":"Basic X..Zw==","index.docker.io":"Basic a...4"}`. Now, it has a valid token and everything worked fine.

**Comment by natalieparellano** [comment]
https://github.com/buildpacks/lifecycle/pull/854 would address the other issue.

**Comment by jromero** [comment]
@natalieparellano is this issue now resolved by https://github.com/buildpacks/lifecycle/pull/862
